src/leo_s3_auth.erl

```diff
@@ -359,6 +455,65 @@ setup(DB, Provider) ->
     true = ets:insert(?AUTH_INFO, {1, #auth_info{db       = DB,
                                                  provider = Provider}}),
     ok.
+%% @doc Trim Space From Binary
+%% @private
+trim(<<>>) ->
+    <<>>;
+trim(Bin) ->
+    trim(Bin, byte_size(Bin)).
+
+trim(Bin = <<Byte:1/binary, Tail/binary>>, Size) ->
+    case is_space(Byte) of
+        true ->
+            trim(Tail, Size - 1);
+        false ->
+            trim_tail(Bin, Size)
+    end.
+
+trim_tail(Bin, Size) ->
+    SizeMinus1 = Size - 1,
+    <<Rest:SizeMinus1/binary, Byte:1/binary>> = Bin,
+    case is_space(Byte) of
+        true ->
+            trim_tail(Rest, Size - 1);
+        false ->
+            Bin
+    end.
+
+is_space(<<" ">>) ->
+    true;
+is_space(_) ->
+    false.
```

Comment:
There are some implementations to trim binary with Erlang.
I shared the result with you.

https://gist.github.com/mocchira/0b637ac0d66389492363

The second one is slightly faster than yours in my env(ec2, 8GB, 4core).
Need less to say, The third (regular expression version) is obviously slow.